src/openrct2/network/Http.cpp

```diff
@@ -14,313 +14,234 @@
  *****************************************************************************/
 #pragma endregion
 
+#include <exception>
+#include <memory>
 #include <thread>
 
-#include "http.h"
+#include "Http.h"
 
-#ifdef DISABLE_HTTP
+#ifndef DISABLE_HTTP
 
-void http_init() { }
-void http_dispose() { }
-
-#else
-
-#include "../core/Console.hpp"
-#include "../core/Math.hpp"
 #include "../Version.h"
+#include "../core/Console.hpp"
 
 #ifdef _WIN32
-    // cURL includes windows.h, but we don't need all of it.
-    #define WIN32_LEAN_AND_MEAN
+// cURL includes windows.h, but we don't need all of it.
+#define WIN32_LEAN_AND_MEAN
 #endif
 #include <curl/curl.h>
 
-#define MIME_TYPE_APPLICATION_JSON "application/json"
 #define OPENRCT2_USER_AGENT "OpenRCT2/" OPENRCT2_VERSION
 
-struct HttpRequest2
+namespace http
 {
-    void *          Tag = nullptr;
-    std::string     Method;
-    std::string     Url;
-    http_data_type  Type;
-    size_t          Size = 0;
-    union
-    {
-        char *      Buffer = nullptr;
-        json_t *    Json;
-    } Body;
 
-    HttpRequest2() { }
+std::atomic<uint8> Client::refcount{ 0 };
 
-    HttpRequest2(const HttpRequest2 &request)
-    {
-        Tag = request.Tag;
-        Method = request.Method;
-        Url = request.Url;
-        Type = request.Type;
-        Size = request.Size;
-        if (request.Type == HTTP_DATA_JSON)
-        {
-            Body.Json = json_deep_copy(request.Body.Json);
-        }
-        else
-        {
-            Body.Buffer = new char[request.Size];
-            memcpy(Body.Buffer, request.Body.Buffer, request.Size);
-        }
-    }
+Client::Client()
+{
+    if (++refcount == 1)
+        curl_global_init(CURL_GLOBAL_DEFAULT);
+}
 
-    explicit HttpRequest2(const http_request_t * request)
-    {
-        Tag = request->tag;
-        Method = std::string(request->method);
-        Url = std::string(request->url);
-        Type = request->type;
-        Size = request->size;
-        if (request->type == HTTP_DATA_JSON)
-        {
-            Body.Json = json_deep_copy(request->root);
-        }
-        else
-        {
-            Body.Buffer = new char[request->size];
-            memcpy(Body.Buffer, request->body, request->size);
-        }
-    }
+Client::~Client()
+{
+    if (--refcount == 0)
+        curl_global_cleanup();
+}
 
-    ~HttpRequest2()
-    {
-        if (Type == HTTP_DATA_JSON)
-        {
-            json_decref(Body.Json);
-        }
-        else
-        {
-            delete Body.Buffer;
-        }
-    }
-};
+static size_t writeData(char * ptr, size_t size, size_t nmemb, void * userdata)
```

Comment:
`ptr` should be `const` and while at it maybe call it `src`?

Reply:
I've copied that straight out of the examples, if I've confirmed that everything builds with const char within function signature I'm fine.